lib/trivial.nix

```diff
@@ -188,6 +188,26 @@ rec {
   warn = msg: builtins.trace "WARNING: ${msg}";
   info = msg: builtins.trace "INFO: ${msg}";
 
+  /* Print a trace message if pred is false.
+     Intended to be used to augment asserts with helpful error messages.
+
+     Example:
+       assertMsg false "nope"
+       => false
+       stderr> trace: nope
+
+       assert (assertMsg ("foo" == "bar") "foo is not bar, silly"); ""
+       stderr> trace: foo is not bar, silly
+       stderr> assert failed at …
+
+     Type:
+       assertMsg :: Bool -> String -> Bool
+  */
+  assertMsg = pred: msg:
+    if pred
+    then true
+    else builtins.trace msg false;
```

Comment:
Why not make the `assert` part of the function?

Reply:
Hm, it has to do with the fact that `assert` is not really an expression (it doesn’t return a value) but a statement (that influences the state of evaluation) of sorts.

We could change it to include an assert, but that means it can’t appear in statement position any longer. Maybe that’s not bad, though.

Reply:
```
let
  assertMsgIf = pred: msg: v: if pred v then v else builtins.trace msg; assert false; /*never reached*/ {};
  assertMsg = pred: assertMsgIf (const pred)
in
  (assertMsg (bar != borz)) /*where to put the derivation?*/

Reply:
Oh right, makes sense